flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/schema/CatalogSourceTable.scala

```diff
@@ -164,10 +164,21 @@ class CatalogSourceTable[T](
     } else {
       TableFactoryUtil.findAndCreateTableSource(context)
     }
+
     if (!tableSource.isInstanceOf[StreamTableSource[_]]) {
       throw new TableException("Catalog tables support only "
         + "StreamTableSource and InputFormatTableSource")
     }
+
+    // validate TableSource when it is in batch mode
+    if (!schemaTable.isStreamingMode) {
```

Comment:
Why not add this to `BatchExecTableSourceScanRule`?

Reply:
My initial thought is "fail-fast", throw the exception at ealier stage, just like the above validation which only allows `StreamTableSource` and `InputFormatTableSource`.